src/interop/js-interop/usage.md

```diff
@@ -0,0 +1,368 @@
+  types: `[]` and `[]=`. These are instance members that match the semantics of
+  JS' [property accessors]. In the above example, `array[i]` gets the value in
+  the `i`th slot of `array`, and `array[i]` sets the value in that slot to
+  `i.toJS`. We expose other JS operators through [utility functions] in
+  `dart:js_interop`.
+
+In the [JS types] section, you'll see that there exists a predefined JS type for
+`Array` called `JSArray`. Here, we're writing another interop type for the same
+JS type and just *viewing* it differently, so there is no conflict. If we
+wanted, we could have made the representation type `JSArray`:
+
+```dart
+extension type Array._(JSArray<JSAny?> _) implements JSArray<JSAny?> {}
+```
+
+Lastly, like any other extension type, you're allowed to declare any
+non-`external` members in the interop type. `isZeroLength` is one such example.
+
+#### Object literal constructors
+
+It is useful sometimes to create a JS [object literal] that simply contains a
+a number of properties and their values. In order to do this, we allow you to
+use a constructor with only named arguments:
+
+```dart
+import ‘dart:js_interop’;
+
+extension type Options._(JSObject o) {
+  external Options({int a, int b});
+  external int get a;
+  external int get b;
+}
+```
+
+A call to `Options(a: 0, b: 1)` will result in a JS invocation of
+`{a: 0, b: 1}`. You can get or set these values through `external` instance
+members.
+
+### Extension members on interop types
+
+You can also write `external` members in extensions of interop types. For
+example:
+
+```dart
+extension on Array {
+  external int push(JSAny? any);
+}
+```
+
+The semantics of calling `push` are identical to what it would have been if it
+was in the definition of `Array` instead. The only `external` members you can
+write in such an extension are instance members and the allowed operators. Like
+with interop types, you can write any non-`external` members in the extension.
+These extensions are useful for when an interop type doesn't expose the
+`external` member you need and you don't want to create a new interop type.
+
+### Parameters
+
+`external` interop methods can only contain positional and optional arguments.
+The one exception is object literal constructors, where they can contain *only*
+named arguments.
+
+Unlike with non-`external` methods, optional arguments do not get replaced with
+their default value, but are instead omitted. For example:
+
+```dart
+external int push(JSAny? any, [JSAny? any2]);
+```
+
+Calling `array.push(0.toJS)` in Dart will result in a JS invocation of
+`array.push(0.toJS)` and *not* `array.push(0.toJS, null)`. If you declare a
+parameter with an explicit default value, you will get a warning that the value
+will be ignored.
+
+## `@JS()`
+
+It is sometimes useful to refer to a JS property with a different name than the
+one written. For example, if you want to write two `external` APIs that point to
```

Comment:
not sure if we want to highlight it, but other common uses for renaming is when the JS name is invalid or unsable in Dart (e.g. a name that starts with `_`, a name that matches a Dart keyword)